actions/src/sema4ai/actions/agent/__init__.py

```diff
@@ -8,9 +8,75 @@
 
 if TYPE_CHECKING:
     from sema4ai.actions import Table
+
+
+# Cache PyArrow availability to avoid repeated imports
+_pyarrow_available = None
+
+def _is_pyarrow_available() -> bool:
+    """Check if PyArrow is available, caching the result."""
+    global _pyarrow_available
+    if _pyarrow_available is None:
+        try:
+            import pyarrow as pa
+            import pyarrow.parquet as pq
+            _pyarrow_available = True
+        except ImportError:
+            _pyarrow_available = False
+    return _pyarrow_available
+
+
+def _parse_dataframe_response(response) -> dict:
+    """Parse dataframe response, trying Parquet first, then falling back to JSON.
+    
+    Args:
+        response: HTTP response object
+        
+    Returns:
+        dict: Parsed dataframe data with columns, rows, name, description
+    """
+    import json
+    
+    # Check if response is Parquet format and PyArrow is available
+    content_type = response.headers.get('content-type', '').lower()
+    if ('parquet' in content_type or 'application/octet-stream' in content_type) and _is_pyarrow_available():
```

Comment:
It should not need to check the content type as you should know whether the request was for parquet or json already.